### DISCUSSION TITLE
Cannot install on go

### DISCUSSION THREAD

**Discussion by mfsiat** [rabbitmq/amqp091-go#132] [Q&A]
I just can't install it over `**go install github.com/rabbitmq/amqp091-go**`
It shows 
```
package github.com/rabbitmq/amqp091-go is not a main package
```
Can anyone help ?

**Comment by Zerpet** [comment]
This is a library and it does not provide a `main` package to run. Essentially, what `go install` is trying to convey 😄 To import this library into your project, you have to use `go get github.com/rabbitmq/amqp091-go` ([reference](https://go.dev/ref/mod#go-get)). You would use `go install` to install packages that provide a binary to run, for example, [mockgen](https://github.com/golang/mock).

**Reply by doodi-v1** [reply]
I think the OP is experiencing the same problem I am, "go get github.com/rabbitmq/amqp091-go", which is shown in various tutorials, returns the following with the latest go version:

go: go.mod file not found in current directory or any parent directory.
        'go get' is no longer supported outside a module.
        To build and install a command, use 'go install' with a version,
        like 'go install example.com/cmd@latest'
        For more information, see https://golang.org/doc/go-get-install-deprecation
        or run 'go help get' or 'go help install'.

**Reply by doodi-v1** [reply]
And "go install github.com/rabbitmq/amqp091-go@latest" returns:

package github.com/rabbitmq/amqp091-go is not a main package

**Reply by lukebakken** [reply]
Hello! Would you mind linking to the specific tutorial causing you issues? I'm assuming it's one of these:

https://github.com/rabbitmq/rabbitmq-tutorials/tree/main/go

My guess is we need to update our examples and tutorials.

Feel free to open a PR in the above repository to bring the tutorial code and instructions up-to-date. We would appreciate it!

**Reply by doodi-v1** [reply]
Yes, that was one of the tutorials I found. The first one I started with was https://x-team.com/blog/set-up-rabbitmq-with-docker-compose/ - which is not yours, but includes the same go get... instruction that I am stuck on. 

It appears the get instruction was deprecated in 1.17 and removed in 1.18 of go. I'm currently uninstalling go and will install an older pre-1.18 version. Not pretty, but I'm just trying to get a quick POC running.

**Reply by Zerpet** [reply]
`go get` is alive and working in 1.19 and 1.20. I believe the key to your problem is in the very first line of the error message:

```
go: go.mod file not found in current directory or any parent directory.
```

In order for `go get` to fetch a module, you must have a `go.mod` file in your current directory, or in a parent directory. It's not specific to this library, but to writing Go code in general, see: https://go.dev/doc/tutorial/getting-started#code

I'm confident that if you [clone the Go tutorial](https://github.com/rabbitmq/rabbitmq-tutorials/tree/main/go), `cd` into the `go` folder and run your `go get` command, it will work without any issue.

To resolve your immediate problem at hand, you have to `cd` into the folder where you have your Go code and run `go mod init example/hello` (feel free to change `example/hello` by a relevant module name to you).

P.s: Go modules are hard, have been hard since Go 1.11 🥲

**Reply by doodi-v1** [reply]
Thanks. I uninstalled the latest go and installed the 1.17 version and it worked without issue. When I get a chance I'll uninstall go again and reinstall the latest to give this a try. I'll let everyone here know the results.

**Reply by doodi-v1** [reply]
Awesome! Simply ran `go mod init example/hello` in my go clone folder, then followed by `go get github.com/rabbitmq/amqp091-go` - worked like a charm. Thanks. Maybe simply add a note in the docs for go ignoramuses like me. 👌
